tests/test_decoder.py

```diff
@@ -0,0 +1,482 @@
+        self.assertTrue(uses_bom('utf_16_be_sig'))
+        self.assertFalse(uses_bom('utf_16_le'))
+        self.assertTrue(uses_bom('utf_16_le_sig'))
+
+        self.assertFalse(uses_bom('utf_8'))
+        self.assertTrue(uses_bom('utf_8_sig'))
+
+    def test_encoding_detectable(self):
+        # No BOM + nothing
+        self.assertFalse(encoding_detectable('', 'utf_32_be'))
+        self.assertFalse(encoding_detectable('', 'utf_32_le'))
+
+        self.assertFalse(encoding_detectable('', 'utf_16_be'))
+        self.assertFalse(encoding_detectable('', 'utf_16_le'))
+
+        self.assertFalse(encoding_detectable('', 'utf_8'))
+        # BOM + nothing
+        self.assertTrue(encoding_detectable('', 'utf_32'))
+        self.assertTrue(encoding_detectable('', 'utf_32_be_sig'))
+        self.assertTrue(encoding_detectable('', 'utf_32_le_sig'))
+
+        self.assertTrue(encoding_detectable('', 'utf_16'))
+        self.assertTrue(encoding_detectable('', 'utf_16_be_sig'))
+        self.assertTrue(encoding_detectable('', 'utf_16_le_sig'))
+
+        self.assertTrue(encoding_detectable('', 'utf_8_sig'))
+        # No BOM + non-ASCII
+        self.assertFalse(encoding_detectable('Ⓝⓔ', 'utf_32_be'))
+        self.assertFalse(encoding_detectable('ⓥⓔ', 'utf_32_le'))
+
+        self.assertFalse(encoding_detectable('ⓡ ', 'utf_16_be'))
+        self.assertFalse(encoding_detectable('ⓖⓞ', 'utf_16_le'))
+
+        self.assertFalse(encoding_detectable('ⓝⓝ', 'utf_8'))
+        # No BOM + ASCII
+        self.assertTrue(encoding_detectable('a ', 'utf_32_be'))
+        self.assertTrue(encoding_detectable('gi', 'utf_32_le'))
+
+        self.assertTrue(encoding_detectable('ve', 'utf_16_be'))
+        self.assertTrue(encoding_detectable(' y', 'utf_16_le'))
+
+        self.assertTrue(encoding_detectable('ou', 'utf_8'))
+        # BOM + non-ASCII
+        self.assertTrue(encoding_detectable('␣ⓤ', 'utf_32'))
+        self.assertTrue(encoding_detectable('ⓟ␤', 'utf_32_be_sig'))
+        self.assertTrue(encoding_detectable('Ⓝⓔ', 'utf_32_le_sig'))
+
+        self.assertTrue(encoding_detectable('ⓥⓔ', 'utf_16'))
+        self.assertTrue(encoding_detectable('ⓡ␣', 'utf_16_be_sig'))
+        self.assertTrue(encoding_detectable('ⓖⓞ', 'utf_16_le_sig'))
+
+        self.assertTrue(encoding_detectable('ⓝⓝ', 'utf_8_sig'))
+        # BOM + ASCII
+        self.assertTrue(encoding_detectable('a ', 'utf_32'))
+        self.assertTrue(encoding_detectable('le', 'utf_32_be_sig'))
+        self.assertTrue(encoding_detectable('t ', 'utf_32_le_sig'))
+
+        self.assertTrue(encoding_detectable('yo', 'utf_16'))
+        self.assertTrue(encoding_detectable('u ', 'utf_16_be_sig'))
+        self.assertTrue(encoding_detectable('do', 'utf_16_le_sig'))
+
+        self.assertTrue(encoding_detectable('wn', 'utf_8_sig'))
+
+
+class DecoderTestCase(unittest.TestCase):
+    def detect_encoding_test_helper(
+        self,
+        original_string,
+        input_bytes,
+        expected_output
+    ):
```

Comment:
These variable names confused me for 2 minutes (I thought "output" referred to a string) :sweat_smile: 
Can you get rid of `original_string` (it isn't used) and use something like this:
```suggestion
    def detect_encoding_test_helper(self, input_bytes, expected_codec):
```
… and 3 lines later `actual_output` → `actual_codec`?

PS: For consistency, we could also change `auto_decode_test_helper()` to use `expected_string` instead of `expected_output`. It would be more explicit and easier to grasp.

Reply:
> Can you get rid of `original_string` (it isn't used) and use something like this:
> […]
> … and 3 lines later `actual_output` → `actual_codec`?

Done.

> PS: For consistency, we could also change `auto_decode_test_helper()` to use `expected_string` instead of `expected_output`. It would be more explicit and easier to grasp.

Done.